### ISSUE TITLE
Баги в 3.0.0

### ISSUE THREAD

**Issue by dvoyakiy** [Mishiranu/Dashchan#240]
Пропала кнопка загрузки в видео
![Screenshot_2020-09-17-23-56-17-977_com.mishiranu.dashchan.png](https://user-images.githubusercontent.com/40234487/93527667-e184f900-f941-11ea-8011-1d97d8a99f14.png)

При загрузке картинки показывает оповещение "необходимо подтверждение" и лоадер. Раньше просто загружало. При этом когда загрузка начинается, не даёт свернуть изображение кнопкой "назад" на телефоне до окончания загрузки. 
![Screenshot_2020-09-17-23-56-55-934_com.mishiranu.dashchan.png](https://user-images.githubusercontent.com/40234487/93527678-e649ad00-f941-11ea-9929-e50027046a30.png)
![Screenshot_2020-09-17-23-56-58-078_com.mishiranu.dashchan.png](https://user-images.githubusercontent.com/40234487/93527702-ec3f8e00-f941-11ea-92fc-243539883cb4.png)

**Comment by Mishiranu** [comment]
А про что последние 2 скриншота?

**Comment by dvoyakiy** [comment]
> А про что последние 2 скриншота?

> При загрузке картинки показывает оповещение "необходимо подтверждение" и лоадер. Раньше просто загружало. При этом когда загрузка начинается, не даёт свернуть изображение кнопкой "назад" на телефоне до окончания загрузки.

Лоадер с надписью "обработка данных"

**Comment by Mishiranu** [comment]
Теперь вижу. Насколько долго это окно висит, интересно?

**Comment by dvoyakiy** [comment]
> Насколько долго это окно висит, интересно?

До начала загрузки, секунды 2-3

**Comment by Mishiranu** [comment]
Ого. Ну ка, а сколько у тебя файлов в той директории, куда сохранить пытаешься?

**Comment by dvoyakiy** [comment]
> Ого. Ну ка, а сколько у тебя файлов в той директории, куда сохранить пытаешься?

Много, больше 11к. Но суть в том что до апдейта при нажатии на кнопку оно просто в фоне сохраняло (хоть и был полсекундный затуп, но при этом никаких окон и оповещений не было)

**Comment by Mishiranu** [comment]
Да, теперь это по-другому работает. Впрочем, такая длинная задержка вполне объяснима, если в одной директории 11 тысяч файлов. Надо будет придумать, что с этим сделать.

Сделано это лишь для того, чтобы проверить, не существует ли ещё файлов с таким именем, и если существует — предложить перезаписать, сохранить рядом или отменить. Обход 11 тысяч файлов с SAF это довольно трудоёмкая процедура.

**Comment by dvoyakiy** [comment]
> Сделано это лишь для того, чтобы проверить, не существует ли ещё файлов с таким именем, и если существует — предложить перезаписать, сохранить рядом или отменить. Обход 11 тысяч файлов с SAF это довольно трудоёмкая процедура.

А разве раньше этого не было? Но проверялись только недавние файлы или как? А то иногда замечал, что если картинка была загружена давно, то предупреждения что файл существует не было

**Comment by Mishiranu** [comment]
> А разве раньше этого не было?

Раньше я напрямую к файловой системе через обычные обёртки POSIX API обращался, а теперь я использую медленный SAF. Я в треде написал, почему я так делаю. У него есть существенный недостаток в виде пониженной производительности (в несколько десятков раз), из-за этого диалог так долго висит.

> Но проверялись только недавние файлы или как?

Все проверялись.

Вообще я бы посоветовал разбить свои файлы на директории, это гарантированно повысит производительность. Даже со стандартным апи, как ты говоришь, у тебя были зависания, и это на самом деле не окей. Если у тебя андроид ниже 8 (т.е. вместо sdcardfs используется тормозной fuse-драйвер), то там всё совсем плохо будет.

**Comment by IS2511** [comment]
На тему обновления
Страшно обновлятся) Знакомый только что написал, что перестали работать видео вообще. Сидит откатывается на мои версии...

**Comment by Mishiranu** [comment]
Конечно страшно, я там половину кода переписал.

Но вот про то, что видео перестали работать, впервые слышу вообще.

Хотя мне кое-что на почту приходило, там не вполне адекватный баг связанный с видео, я с ним попозже разберусь. Наверное, этот знакомый как раз и писал.

**Comment by Mishiranu** [comment]
Проблема с отсутствием кнопок должна быть исправлена в 3.0.1 (коммит https://github.com/Mishiranu/Dashchan/commit/946063dadf6b12771eaa656fe4b51a5509946548)

**Comment by Vad1us** [comment]
![Screenshot_Настройки_20200918-064547](https://user-images.githubusercontent.com/17219712/93553009-a4892880-f97a-11ea-884c-d083103d83f7.png)
Не могу дать разрешение на доступ к памяти. Пробовал удалить все (проверил осталась ли папка), но ничего не помогло. Даже вручную нельзя дать доступ, в списке тупо нет дашчана. А без доступа не грузит картинки. Версия 3.0.1. андроид 8.1

**Comment by shinji2009** [comment]
То же самое все. Не могу дать разрешение. Та же версия, тот же андроид

**Comment by dvoyakiy** [comment]
> Вообще я бы посоветовал разбить свои файлы на директории

Ок, тогда пока буду сохранять в вообще другую папку.

Ещё кстати заметил, при разворачивании приложения где то секунду висит чёрный экран (или просто обои чёрные делаются) и тогда уже открывается дешчан.

**Comment by LLlOZAHAX** [comment]
На двух Meizu вот такая фигня, когда делаю всё по инструкции.
https://yadi.sk/i/IHIgH5S83KdyWg

**Comment by Mishiranu** [comment]
Полагаю, оно так крашится. Лог будет?

**Comment by LLlOZAHAX** [comment]
> Полагаю, оно так крашится. Лог будет?

[error-1600407131713.txt](https://github.com/Mishiranu/Dashchan/files/5251299/error-1600407131713.txt)

**Comment by Mishiranu** [comment]
Это точно не то, такой краш происходил на 3.0.0 и сейчас не актуален.
Поставь 3.0.2, удали все файлы в директории errors и попробуй снова лог сними.

**Comment by LLlOZAHAX** [comment]
Прошу прощения, но лог не пишется. Обновился на 3.0.2, ситуация с вылетами на приложенном видео. https://yadi.sk/i/CgNk-6XQm_GUGQ

**Comment by Mishiranu** [comment]
Что значит не пишется? Не появляются новые файлы?

**Comment by LLlOZAHAX** [comment]
По рекомендации удалил предыдущий лог - новые не пишутся.

**Comment by Mishiranu** [comment]
Если не пишутся, значит не крашится. А в таком случае я совершенно не понимаю, что происходит. Выглядит как краш, потому что сбрасывается состояние экрана. Могу разве что предложить полный лог системы через adb снять.

**Comment by N0D4N** [comment]
Проблема с долгой загрузкой, блокируещем лоадере и фризящим UI остается, бесит и мешает браузингу даже если в папках  мало файлов (<1к).
Как вариант, можно загружать файлы в фоне, давая юзеру браузить дальше и в случае если окажется что такой файл в папке уже есть - сделать всплывашку, возможно с названием файла и превьюшкой, промтящей юзера на перезапись, пропуск или ренейминг. 
Или же при запуске приложения обходить всю папку с загрузками в фоновом потоке, строить хешсет, которую будет использовать загрузчик, и тогда проверка на наличие файла будет O(1) вместо текущего O(n) которое каждый раз выполняется при каждой загрузке. В таком случае будут проблемы если пользователь что-то наделал с папкой загрузок между запуском приложения и нажатия на кнопку загрузки, но в таком случае можно дефолтить на молчаливую перезапись, молчаливый пропуск или вывод ошибки через нотификации, опционально чтобы поведение это выбиралось в настройках. Ну и соответсвенно хешсет должен обновляться с каждым новым загруженным файлом.
